### PR TITLE
DOC: Enhance asfreq docs

### PR DESCRIPTION
**Pull request by plammens**
- [n/a] closes #xxxxx
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] tests added

Enhanced some docstrings for time-series related methods; mainly `asfreq`. Namely:
 - Converted `NDFrame.asfreq` into a template to use the specific `klass` in `Series.asfreq` and `DataFrame.asfreq`
 - Fully explain the behaviour of `NDFrame.asfreq` in the extended summary.
 - Add an extended summary to `asfreq` in `PeriodArray`/`PeriodIndex`
 - Use `klass` kwarg in docstring template for `PeriodAray.asfreq`\`PeriodIndex.asfreq`
 - Add an extended summary to `date_range`
 - Other very minor fixes/enhancements

The original motivator for this PR was [this question](https://stackoverflow.com/q/63860589/6117426), in particular the confusion of the asker.

### CODE REVIEW

File: pandas/core/arrays/period.py
Comment (MarcoGorelli) [inline pandas/core/arrays/period.py:508]: Thanks @plammens - I haven't tried running this yet, but why are you doubling the brackets here? Single brackets seems consistent with the docstring guide https://pandas.pydata.org/docs/development/contributing_docstring.html

File: pandas/core/generic.py
Comment (MarcoGorelli) [inline pandas/core/generic.py:7538]: Did you mean to write it twice here?

File: pandas/core/generic.py
Comment (MarcoGorelli) [inline pandas/core/generic.py:7543]: No full-stop needed here IMO
```suggestion
        last entries in the original index (see :func:`pandas.date_range`). The
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py
Comment (plammens) [reply pandas/core/generic.py:7538]: Hmm, I don't see it, what do you mean?

Comment (MarcoGorelli) [reply pandas/core/generic.py:7538]:
```
:meth:`PeriodIndex.asfreq pandas.PeriodIndex.asfreq`
```

Is it correct to have `PeriodIndex.asfreq` twice?

Comment (plammens) [reply pandas/core/generic.py:7538]: Oh I see! That's Sphinx syntax, the first part is the link text (`PeriodIndex.asfreq`) and the second part is the actual link path (`pandas.PeriodIndex.asfreq`). I did this just to omit the leading `pandas.` from the link text. (Here `~` wouldn't do it because it would just show `asfreq`.)

Comment (MarcoGorelli) [reply pandas/core/generic.py:7538]: I'm probably missing something, but when I build this locally it shows up twice rather than having a link:
image

Comment (MarcoGorelli) [reply pandas/core/generic.py:7538]: Should it be something like
```suggestion
        :meth:`PeriodIndex.asfreq <pandas.PeriodIndex.asfreq>` (so the original index
```
?

Comment (plammens) [reply pandas/core/generic.py:7538]: Ah, you're totally correct! For some reason I interpreted the angle brackets as not part of the syntax itself—and I didn't notice that when checking the output.

**Comment by plammens** [comment]
@MarcoGorelli  Alright, it should be all done now. I've extracted a generic `call_once` context manager (for the warnings thing) to `pandas.util._call_once`, and added tests for that. The two of the CI builds are failing on a completely unrelated test (`tests.plotting.TestDataFrameGroupByPlots.test_series_groupby_plotting_nominally_works`), and I can't for the life of me figure out why.

**Comment by MarcoGorelli** [comment]
> The two of the CI builds are failing on a completely unrelated test (`tests.plotting.TestDataFrameGroupByPlots.test_series_groupby_plotting_nominally_works`), and I can't for the life of me figure out why.

They're almost certainly unrelated, don't worry about them

Will take another look later, thanks for updating!

**Comment by MarcoGorelli** [comment]
Does this have to all be part of the same PR, or could `check_deprecated_resample_kwargs` be a separate one?

**Comment by plammens** [comment]
> Does this have to all be part of the same PR, or could `check_deprecated_resample_kwargs` be a separate one?

It has to be part of the same PR because of how the warnings were managed previously: the warnings were raised in `Grouper.__new__`, for all possible paths that led there, and the `stacklevel` was set based on whether the call came from a `TimeGrouper` or not (because at the time, every public call of `resample` ended up at `TimeGrouper` in exactly three stack frames).

https://github.com/pandas-dev/pandas/blob/8df0218a4bfae50a7191ff0869e63c28884fb733/pandas/core/groupby/grouper.py#L227-L268

This is very brittle—as this PR demonstrates. When I made the changes to `resample` docs, I made `NDFrame.resample` into a template, to allow for customized `DataFrame.resample` and `Series.resample` docs. This introduces a new stack frame since these methods have to call `super().resample(...)`: 

https://github.com/pandas-dev/pandas/blob/7e8812318d62278712999f6b8351a237f3a6c582/pandas/core/frame.py#L9127-L9156

This then breaks the hacky `stacklevel` munging mentioned above since now for `DataFrame` and `Series` the `stacklevel` should technically have been `5`.

In fact if you run the tests on commit 27b16cf, you will see that they fail on `pandas.tests.resample.test_deprecated`. When I added the `check_resample_deprecated_kwargs` thing in e78a415 to fix this issue.

**Comment by MarcoGorelli** [comment]
Thanks @plammens for explaining - from what I've understood, you want to share docstrings between `DataFrame.resample` and `Series.resample`. However, this causes problems with how the warnings are managed in `Grouper.__new__`.

Would it be possible to make one PR to change how warnings are managed in `Grouper.__new__`? You write

> This then breaks the hacky stacklevel munging mentioned above

so would it be possible to have 1 PR to improve the stacklevel munging and then, once that's in, another PR to share docstrings between `DataFrame.resample` and `Series.resample`? Sorry if I've misunderstood, in which case please do correct me

**Comment by plammens** [comment]
> Thanks @plammens for explaining - from what I've understood, you want to share docstrings between `DataFrame.resample` and `Series.resample`. However, this causes problems with how the warnings are managed in `Grouper.__new__`.

Yes, exactly.

I've made a new PR (#36629), but until that gets merged, if I remove the relevant commits from here, this PR will be failing the tests, so should I wait until the other PR gets merged?

**Comment by MarcoGorelli** [comment]
> I've made a new PR (#36629),

Awesome, thanks

> so should I wait until the other PR gets merged?

Yup

**Comment by plammens** [comment]
> This pull request is stale because it has been open for thirty days with no activity. Please update or respond to this comment if you're still interested in working on this.

Waiting for #36629 (or an equivalent fix) to be merged.

**Comment by jreback** [comment]
> > This pull request is stale because it has been open for thirty days with no activity. Please update or respond to this comment if you're still interested in working on this.
> 
> Waiting for #36629 (or an equivalent fix) to be merged.

why is this a blocker?

**Comment by MarcoGorelli** [comment]
> > > This pull request is stale because it has been open for thirty days with no activity. Please update or respond to this comment if you're still interested in working on this.
> > 
> > 
> > Waiting for #36629 (or an equivalent fix) to be merged.
> 
> why is this a blocker?

As far as I understand, it's because he wants to share docs between `DataFrame.resample` and `Series.resample`, though doing so would change the stacklevel

**Comment by jreback** [comment]
let just fix the appropriate tests (eg turn off the stack level checks)

these aren't that important for a really old warning

**Comment by MarcoGorelli** [comment]
Sure - happy to do that, @plammens ? Sorry for having led you down the wrong path

**Comment by plammens** [comment]
> let just fix the appropriate tests (eg turn off the stack level checks)
> 
> these aren't that important for a really old warning

There is already an open PR for that - #37603 :)

**Comment by plammens** [comment]
After discussion in #37603, it was decided to just update the stacklevel accordingly and disable some of the stacklevel checks for those cases in which the stacklevel couldn't be computed easily (at the same time as the one for `Series.resample` and `DataFrame.resample`). I updated the PR to reflect this.

I rebased onto the latest passing master. There have been lots of changes meanwhile, so it is possible I messed up something while resolving merge conflicts.

CC @jorisvandenbossche

**Review by jreback — APPROVED** [review]
lgtm @jbrockmendel @jorisvandenbossche if comments

File: pandas/core/arrays/period.py
Comment (jbrockmendel) [inline pandas/core/arrays/period.py:75]: should "array" be changed to the same pattern as for period_array?

File: pandas/core/generic.py
Comment (jbrockmendel) [inline pandas/core/generic.py:7734]: NaT?

File: pandas/core/generic.py
Comment (jbrockmendel) [inline pandas/core/generic.py:7780]: `'s':series` -> `"s": series`?

File: pandas/core/generic.py
Comment (jbrockmendel) [inline pandas/core/generic.py:8233]: why was the space removed here?

File: pandas/core/indexes/datetimes.py
Comment (jbrockmendel) [inline pandas/core/indexes/datetimes.py:933]: timesteps -> timestamps?

Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py:933]: time offset -> difference?

File: pandas/core/arrays/period.py
Comment (plammens) [reply pandas/core/arrays/period.py:75]: I assume this is `numpy.array`?

File: pandas/core/generic.py
Comment (plammens) [reply pandas/core/generic.py:7780]: Thanks, I'll add the space; but all of the other examples use single quotes, so I'd stick to that.

File: pandas/core/generic.py
Comment (plammens) [reply pandas/core/generic.py:7734]: I'm not sure what you mean, the behaviour as shown in the examples uses `NaN`?

File: pandas/core/generic.py
Comment (plammens) [reply pandas/core/generic.py:8233]: Probably a typo, thanks.

File: pandas/core/indexes/datetimes.py
Comment (plammens) [reply pandas/core/indexes/datetimes.py:933]: I used "timesteps" here to refer to each of the time points in the time series; maybe this is not standard terminology though I think it's used elsewhere in the docs. Maybe should use "time points"?

File: pandas/core/arrays/period.py
Comment (jbrockmendel) [reply pandas/core/arrays/period.py:75]: im pretty sure this refers to pd.array

File: pandas/core/generic.py
Comment (jbrockmendel) [reply pandas/core/generic.py:7734]: yep, my bad

File: pandas/core/generic.py
Comment (plammens) [reply pandas/core/generic.py]: Fixed in 9c91ee7.

File: pandas/core/generic.py
Comment (plammens) [reply pandas/core/generic.py]: Fixed in 9c91ee7.

File: pandas/core/arrays/period.py
Comment (plammens) [reply pandas/core/arrays/period.py]: Changed in 9c91ee7.

File: pandas/core/indexes/datetimes.py
Comment (plammens) [reply pandas/core/indexes/datetimes.py]: Edited in 9c91ee7, does it look good now?

**Comment by jreback** [comment]
thanks @plammens very nice

**Comment by jreback** [comment]
pls have a look at the built docs (may take a bit): https://pandas.pydata.org/pandas-docs/dev/ and if needed followup

**Pull request merged**
